File: node/fuzz/targets/parse-der.mjs
/**
 * The bounded DER/BER reader, on its own. Every other target reaches it
 * through a structure walk; this one hands it arbitrary bytes directly so a
 * length or depth bug shows up without a CMS or certificate shape around it.
 *
 * The X.509 helpers `der.ts` hand-writes are exercised on the same bytes.
 * They slice the input a second time from offsets parsing chose, which is
 * exactly the state a mutated length can leave wrong. (The certificate
 * *template* is not fuzzed here: the Node build parses certificates with
 * `node:crypto`'s `X509Certificate`, i.e. OpenSSL, not with code from this
 * repository — that is the one Rust target with no Node counterpart.)
 *
 * Invariant: one well-formed value or a `ParseError`, never a `TypeError`
 * or a `RangeError` from indexing past the end of the input.
 */
import {
  Tag,
  hasExtension,
  isOctetString,
  octetStringValue,
  parse,
  tbsParts,
} from '../../dist/der.js';
import { PARSE_ERRORS, requireTypedError } from '../harness.mjs';

// The Apple receipt-signing marker OID, i.e. the extension lookup the
// receipt path actually performs.
const RECEIPT_SIGNER_OID = '1.2.840.113635.100.6.11.1';

export function fuzz(data) {
  let node;
  try {
    node = parse(data);
  } catch (error) {
    requireTypedError(error, 'parse', PARSE_ERRORS);
    return;
  }
  try {
    if (isOctetString(node)) {
      octetStringValue(node);
    }
    if (node.tag === Tag.SEQUENCE) {
      tbsParts(data);
      hasExtension(data, RECEIPT_SIGNER_OID);
    }
  } catch (error) {
    requireTypedError(error, 'der accessors', PARSE_ERRORS);
  }
}

File: ruby/fuzz/targets/parse_der.rb
# frozen_string_literal: true

# The bounded ASN.1 reader, on its own. Every other target reaches it through
# a structure walk; this one hands it arbitrary bytes so a length or depth bug
# shows up without a CMS or certificate shape around it.
#
# The invariant is the module's own documented contract: one well-formed value
# or an Asn1::Error. A SystemStackError here would mean the iterative scanner
# had grown a recursive path again, which is the exact regression
# Asn1.scan! exists to prevent.

require "ruzzy"
require_relative "../support"

APRV = FuzzSupport::APRV

TEST_ONE_INPUT = lambda do |data|
  outcome, = FuzzSupport.call("Asn1.scan!", APRV::Asn1::Error) { APRV::Asn1.scan!(data) }
  # The tree builder only ever runs on bytes the scanner passed, so that is
  # the only shape it is fuzzed in.
  FuzzSupport.call("Asn1.parse", APRV::Asn1::Error) { APRV::Asn1.parse(data) } if outcome == :accepted
  nil
end

Ruzzy.fuzz(TEST_ONE_INPUT)
